### PR TITLE
doc: Update mainpage

### PR DESCRIPTION
**Pull request by authmillenon**
To give an idea how the [text currently authored in an etherpad](https://padlite.spline.inf.fu-berlin.de/p/mjtL88j5np) will be incorporated, I decided to open a preliminary Pull Request already.

~~Depends on #4090.~~

### CODE REVIEW

**Comment by authmillenon** [comment]
Wait... For some reason this ruins sectioning in the sidebar (the chapters on the main page are now under "Modules" the Oo). Will try to find a fix.

**Comment by authmillenon** [comment]
Updated to current version of the pad (and spellchecked that ;-)).

**Comment by authmillenon** [comment]
Removed overview graphic for now. Will make #4090 dependent on this PR (so reverse the dependencies)

**Comment by haukepetersen** [comment]
looks good as a starting point from my side, after this is merged it will be easy to build upon it. @OlegHahm, @emmanuelsearch: do you ack this for now?

File: doc/doxygen/src/mainpage.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+output after running `make doc`.
+
+Lastly, the `dist` folder contains tools to help you with RIOT. These include
+the serial terminal application `pyterm`, generic scripts for flashing,
+debugging, reseting (e.g. support for [OpenOCD](http://openocd.org/),
+[Jlink](https://www.segger.com/jlink_base.html)), as well as code enabling easy
+integration to open testbeds such as the [IoT-LAB](https://www.iot-lab.info/).
+Furthermore you can find here scripts to do all kind of code and style checks.
+
+Further information                                      {#further-information}
```
Comment (OlegHahm) [inline doc/doxygen/src/mainpage.md]: Shouldn't this be commented out until we have content?

File: doc/doxygen/src/mainpage.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+devices. These requirements include a very low memory footprint (on the order
+of a few kilobytes), high energy efficiency, real-time capabilities,
+communication stacks for both wireless and wired networks, and support for a
+wide range of low-power hardware.
+
+RIOT provides a microkernel, multiple network stacks, and utilities which
+include cryptographic libraries, data structures (bloom filters, hash tables,
+priority queues), a shell and more. RIOT supports a wide range of
+microcontroller architectures, radio drivers, sensors, and configurations for
+entire platforms, e.g. ATSAMR21, Zolertia Z1, STM32 Discovery Boards etc. (see
```
Comment (OlegHahm) [inline doc/doxygen/src/mainpage.md]: `s/ATSAMR21/Atmel SAM R21 Xplained Pro/`?

File: doc/doxygen/src/mainpage.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+structure, you will easiy find your way around in RIOT's code base.
+
+RIOT's code base is structured into five groups.
+
+<!-- TODO: add graphic -->
+
+ - The kernel (`core`)
+ - Platform specific code (`cpu`; `boards`)
+ - Device drivers (`drivers`)
+ - Device drivers (`drivers`)
```
Comment (OlegHahm) [inline doc/doxygen/src/mainpage.md]: duplicate

File: doc/doxygen/src/mainpage.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+with RIOT, see [TODO: link to getting started guide].
+
+Before that, skimming through the next section is recommended (but not
+mandatory).
+
+Structure                                                          {#structure}
+=========
+
+This section walks you through RIOT's structure. Once you understand this
+structure, you will easiy find your way around in RIOT's code base.
```
Comment (OlegHahm) [inline doc/doxygen/src/mainpage.md]: `s/easiy/easily/`

File: doc/doxygen/src/mainpage.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+ - Device drivers (`drivers`)
+ - Device drivers (`drivers`)
+ - Libraries and network code (`sys/pkg`)
+ - Applications for demonstrating features and for testing (`examples`;
+   `tests`)
+
+In addition RIOT includes a collection of scripts for various tasks (`dist`) as
+well as a predefined environment for generating this documentation (`doc`).
+
+These groups are projected onto the folder structure of RIOT, where each of
```
Comment (OlegHahm) [inline doc/doxygen/src/mainpage.md]: Which groups?

File: doc/doxygen/src/mainpage.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+The board part contains the specific configuration for the CPU it contains.
+This configuration mainly includes the peripheral configuration and
+pin-mapping, the configuration of on-board devices, and the CPU's clock
+configuration.
+
+On top of the source and header files needed for each board, this folder
+additionally may include some script and configuration files needed for
+interfacing with the board. These are typically custom flash/debug scripts or
+e.g. OpenOCD configuration files. For most boards, these files are located in a
+`dist` sub-directory of the board.
```
Comment (OlegHahm) [inline doc/doxygen/src/mainpage.md]: I remember that @LudwigKnuepfer once advocated to use the term "directory" over "folder" (see [Folder metaphor](https://en.wikipedia.org/wiki/Directory_%28computing%29#Folder_metaphor)). I don't think that it is of high importance to be dogmatic here, but I still would prefer not to mix the terms "folder" and "directory" in one document.

File: doc/doxygen/src/mainpage.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+----
+This folder contains the actual kernel. The kernel consists of the scheduler,
+inter-process-communication (messaging), threading, thread synchronization, and
+supporting data-structures and type definitions.
+
+See @ref core for further information and API documentations.
+
+boards
+------
+The platform dependent code is split into two logic elements: CPU's and boards,
```
Comment (OlegHahm) [inline doc/doxygen/src/mainpage.md]: No apostrophe after CPU is required here.

File: doc/doxygen/src/mainpage.md [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+startup code, clock inialization code and thread handling (e.g. context
+switching) code.  For most CPU's you will also find the linkerscripts in the
+`ldscripts`
+sub-directory.
+
+In the `periph` sub-directory of each CPU you can find the implementations of
+the CPU's peripheral drivers like SPI, UART, GPIO, etc.  See @ref driver_periph
+for their API documentation.
+
+Many CPU's share a certain amount of their code (e.g. all ARM Cortex-M based
```
Comment (OlegHahm) [inline doc/doxygen/src/mainpage.md]: s/CPU's/CPUs/`

File: doc/doxygen/src/mainpage.md [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+For each supported CPU this folder contains a sub-directory with the name of
+the CPU. These folders then contain all CPU specific configurations, such as
+implemenations of power management (LPM), interrupt handling and vectors,
+startup code, clock inialization code and thread handling (e.g. context
+switching) code.  For most CPU's you will also find the linkerscripts in the
+`ldscripts`
+sub-directory.
+
+In the `periph` sub-directory of each CPU you can find the implementations of
+the CPU's peripheral drivers like SPI, UART, GPIO, etc.  See @ref driver_periph
```
Comment (OlegHahm) [inline doc/doxygen/src/mainpage.md]: Shouldn't it be rather `CPUs'` here?

File: doc/doxygen/src/mainpage.md [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+See @ref net for more details on networking code.
+
+pkg
+---
+RIOT comes with support for a number of external libraries (e.g.
+[OpenWSN](http://www.openwsn.org/),
+[microcoap](https://github.com/1248/microcoap)). The way they are included is
+that RIOT ships with a custom Makefile for each supported library that
+downloads the library and optionally applies a number of patches to make it
+work with RIOT. These Makefiles and patches you find in the `pkg` folder.
```
Comment (OlegHahm) [inline doc/doxygen/src/mainpage.md]: `s/you find/can be found/` reads IMO less clumsy.

File: doc/doxygen/src/mainpage.md [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+Here you find a number of example applications that demonstrate certain
+features of RIOT. The default example found in this folder is a good starting
+point for anyone who is new to RIOT.
+
+For more information best browse that folder and have a look at the `README.md`
+files that ship with each example.
+
+tests
+-----
+Many features/module in RIOT come with their own test application, which are
```
Comment (OlegHahm) [inline doc/doxygen/src/mainpage.md]: `s/module/modules/`

File: doc/doxygen/src/mainpage.md (hunk 6 above)
Comment (authmillenon) [inline doc/doxygen/src/mainpage.md]: I also prefer folder over directory. Will fix.

File: doc/doxygen/src/mainpage.md (hunk 5 above)
Comment (authmillenon) [inline doc/doxygen/src/mainpage.md]: The groups stated in the two paragraphs above.

File: doc/doxygen/src/mainpage.md (hunk 6 above)
Comment (OlegHahm) [inline doc/doxygen/src/mainpage.md]: Eh, IIRC @LudwigKnuepfer voted the other way around. (Please correct me if I'm wrong, Ludwig.)

File: doc/doxygen/src/mainpage.md (hunk 9 above)
Comment (authmillenon) [inline doc/doxygen/src/mainpage.md]: > for each CPU there is the implementation of the peripheral drivers of the CPU 

vs

> for each CPU there is the implementation of the peripheral drivers of the CPUs

Logically the first statement makes more sense, so singular genitive makes more sense

**Comment by authmillenon** [comment]
Addressed comments.

File: doc/doxygen/src/mainpage.md (hunk 9 above)
Comment (OlegHahm) [inline doc/doxygen/src/mainpage.md]: true

**Comment by OlegHahm** [comment]
ACK

**Comment by authmillenon** [comment]
Squashed

**Comment by emmanuelsearch** [comment]
Are we not missing a graphical element in this page? I thought we had agreed to also add something like https://github.com/authmillenon/RIOT/blob/doc/enh/overview-graphic/doc/doxygen/src/riot-structure.svg

**Comment by authmillenon** [comment]
@emmanuelsearch see #4090

**Pull request merged**